packages/contracts/contracts/SmartInvoiceSplitEscrow.sol

```diff
@@ -0,0 +1,587 @@
+            IERC20(token).safeTransfer(provider, providerAmount);
+            released = released + amount;
+            unchecked {
+                _remaining = _remaining - 1;
+            }
+            emit Release(_milestone, amount);
+        } else {
+            require(balance > 0, "balance is 0");
+
+            uint256 daoAmount = (balance * daoFee) / 10000;
+            uint256 providerAmount = balance - daoAmount;
+            IERC20(token).safeTransfer(dao, daoAmount);
+            IERC20(token).safeTransfer(provider, providerAmount);
+            released = released + balance;
+            emit Release(_milestone, balance);
+        }
+    }
+
+    /**
+     * @dev Internal function to release funds from the contract to the provider.
+     */
+    function _release() internal {
+        // client transfers locker milestone funds to provider and dao
+        uint256 currentMilestone = amounts.length - 1;
+        for (uint256 i = 0; i < amounts.length; i++) {
+            if (milestoneReleased[i] == false) {
+                currentMilestone = i;
+                break;
+            }
+        }
+        _release(currentMilestone);
+    }
+
+    /**
+     * @dev External function to release funds from the contract to the provider.
+     * Uses the internal `_release` function to perform the actual release.
+     */
+    function release() external override nonReentrant {
+        return _release();
+    }
+
+    /** @dev External function to release funds from the contract to the provider.
+     * Uses the internal `_release` function to perform the actual release.
+     * @param _milestone The milestone to release funds to
+     */
+    function release(uint256 _milestone) external override nonReentrant {
+        // client transfers locker funds upto certain milestone to provider and dao
+        require(_milestone < amounts.length, "invalid milestone");
+        require(!milestoneReleased[_milestone], "milestone already released");
+        _release(_milestone);
+    }
+
+    /** @dev External function to release funds from the contract to the provider.
+     * Uses the internal `_release` function to perform the actual release.
+     * @param _milestones The milestones to release funds
+     */
+    function release(uint256[] calldata _milestones) external nonReentrant {
+        // client transfers locker funds upto certain milestone to provider and dao
+        require(!locked, "locked");
+        require(_msgSender() == client, "!client");
+        // require(_milestones >= milestone, "milestone passed");
+        for (uint256 i = 0; i < _milestones.length; i++) {
+            require(_milestones[i] < amounts.length, "invalid milestone");
+            require(
+                !milestoneReleased[_milestones[i]],
+                "milestone already released"
+            );
+        }
+
+        uint256 balance = IERC20(token).balanceOf(address(this));
+        uint256 amount = 0;
```

Comment:
FYI its not necessary to instantiate vars explicitly at 0. All vars are automatically initiated at the null value for their type, eg `uint256 amount;` is the same thing as `uint356 amount = 0;`

Reply:
also, you're looping through the same array twice. If you do the two checks on lines 354 and 355 as part of the loop starting on 365, you can get rid of that first loop.